flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcDynamicTableSource.java

```diff
@@ -96,21 +96,54 @@ public JdbcDynamicTableSource(
     public LookupRuntimeProvider getLookupRuntimeProvider(LookupContext context) {
         // JDBC only support non-nested look up keys
         String[] keyNames = new String[context.getKeys().length];
-        for (int i = 0; i < keyNames.length; i++) {
+
+        for (int i = 0; i < context.getKeys().length; i++) {
             int[] innerKeyArr = context.getKeys()[i];
             Preconditions.checkArgument(
                     innerKeyArr.length == 1, "JDBC only support non-nested look up keys");
             keyNames[i] = DataType.getFieldNames(physicalRowDataType).get(innerKeyArr[0]);
         }
+
         final RowType rowType = (RowType) physicalRowDataType.getLogicalType();
+
+        String[] conditions = null;
+
+        if (this.resolvedPredicates != null) {
+            conditions = new String[this.resolvedPredicates.size()];
+            for (int i = 0; i < this.resolvedPredicates.size(); i++) {
+                String resolvedPredicate = this.resolvedPredicates.get(i);
+                String param = this.pushdownParams[i].toString();
+                /*
+                 * This replace seems like it should be using a Flink class to resolve the parameter. It does not
+                 * effect the dialects as the placeholder comes from JdbcFilterPushdownPreparedStatementVisitor.
+                 *
+                 * Here is what has been considered as alternatives.
+                 *
+                 * We cannot use the way this is done in getScanRuntimeProvider, as the index we have is the index
+                 * into the filters, but it needs the index into the fields. For example one lookup key and one filter
+                 * would both have an index of 0, which the subsequent code would incorrectly resolve to the first
+                 * field.
+                 * We cannot use the PreparedStatement as we have not got access to the statement here.
+                 * We cannot use ParameterizedPredicate as it takes the filter expression as input (e.g EQUALS(...)
+                 * not the form we have here an example would be ('field1'= ?).
+                 */
+                conditions[i] =
+                        (resolvedPredicate.replace(
+                                JdbcFilterPushdownPreparedStatementVisitor
+                                        .PUSHDOWN_PREDICATE_PLACEHOLDER,
+                                param));
```

Comment:
how will it behave if I have `WHERE` condition like `a = 'X?Y'`
or with double quotes it's possible to set special character in column names, so what if there is a column name containing `?` like `"column?" = '??'`

Reply:
coincidentally I was just thinking about that.  As we only support simple expressions - I suspect it should work by replacing the last `?`. I will do some testing - and see what I can make break.

Reply:
@snuyanzin I did some testing, it was not obvious that ? was handled as a wild card for every database, for example Derby will accept LIKE "AA%" but not LIKE "AA?". But I see from Google that other databases do support it.